### PR TITLE
Add formatting check action and apply an initial format

### PR DESCRIPTION
**Pull request by t-bltg**
This PR is a proposition in order to add consistency across the source and test files (after https://github.com/JuliaIO/HDF5.jl/pull/983#issuecomment-1175011022).

The tests are hard to read because the `@testset` blocks are not indented: https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/test/plain.jl#L438
requiring (spurious) manual addition of comment such as `end # testset "Test h5d_fill`.

The source files mixes both `function ... end` and one liner function declarations like :
https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/datasets.jl#L320
, and https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/datasets.jl#L332
making this hard to follow (inconsistent).

The declarations like:
https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/groups.jl#L115
are hard to read and it would be IMO better if they could be split into multiple lines such as https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/readwrite.jl#L247
(basically break on a specific margin / line width).

The indentation is inconsistent e.g. https://github.com/JuliaIO/HDF5.jl/blob/d6ecf42d035784f78616a24b03cda7c037eda7c9/src/api/helpers.jl#L299 which uses 3 spaces instead of 4 (https://docs.julialang.org/en/v1/manual/style-guide/).

I'll stop there with the examples, my point being that I don't care which style convention is used (this is very opinionated), **as long as it is consistent across all source files**.

Currently in this repo, no style guidelines are provided for developers, inducing wasted time on style discussions (e.g. https://github.com/JuliaIO/HDF5.jl/pull/983#issuecomment-1175011022 or https://github.com/JuliaIO/HDF5.jl/pull/917#discussion_r830190912).

If needed, formatting auto-generated files can be omitted  by adding `#! format: off` to a file header (e.g. `gen/gen_wrappers.jl` -> `src/api/functions.jl`) ?

I propose that we only apply formatting to `src` and `test` directories (for the end users), but this could be extended to other directories like `gen` (to be discussed ?).

All the current inconsistencies can be handled in an automated manner using a formatting check action (auto format PRs should be avoided, because they pollute the git history: we should instead use (blocking or non blocking) checks for PRs, see the added `.github/workflows/Format-check.yml` action).

- [ ] discuss which stylistic conventions are to be used (I'd vote for `BlueStyle` + tweaks in `.JuliaFormatter`).

### CODE REVIEW

**Comment by mkitti** [comment]
Available style codifications:

* [JuliaFormatter.BlueStyle](https://domluna.github.io/JuliaFormatter.jl/stable/blue_style/#JuliaFormatter.BlueStyle): https://github.com/invenia/BlueStyle#method-definitions
* [YASGuide](https://domluna.github.io/JuliaFormatter.jl/stable/yas_style/): https://github.com/jrevels/YASGuide
* SciML style: https://github.com/SciML/SciMLStyle

@musm seems to have style thoughts in this comment https://github.com/JuliaIO/HDF5.jl/pull/917#discussion_r830190912. Perhaps he would like to codify them via JuliaFormatter.

**Comment by mkitti** [comment]
To merge this, I would like to try to clear the pull request queue as much as possible. Merging this will require significant work to make those branches work again.

**Comment by t-bltg** [comment]
> Merging this will require significant work to make those branches work again.

Absolutely, this can wait until the queue is low.

**Comment by mkitti** [comment]
I resolved the merge conflict.

**Comment by t-bltg** [comment]
Rebased since the format tests where failing.

**Comment by t-bltg** [comment]
I think it's time to go forward on this one (recent PR queue is low). @musm, what do you think ?

File: src/attributes.jl
Comment (musm) [inline src/attributes.jl:112]: This  shouldn't be a  one liner

Comment (t-bltg) [reply src/attributes.jl:112]: You mean we should manually convert to plain `if ... end`, right ?

**Comment by mkitti** [comment]
For this pull request, we should primarily focus on getting the settings to `.JuliaFormatter.toml` correct. I should be able to pull master, check out `.JuliaFormatter.toml`, apply the formatting, and see almost diff to this branch. Other stylistic changes should be reserved for other pull requests.

**Comment by t-bltg** [comment]
> For this pull request, we should primarily focus on getting the settings to .JuliaFormatter.toml correct. I should be able to pull master, check out .JuliaFormatter.toml, apply the formatting, and see almost diff to this branch.

This should be the case (I've renamed the 2nd commit to show that I only ran `format(["src", "test"])` after creating `.JuliaFormatter` configuration file (1st commit in this PR)).

File: .github/workflows/Format-check.yml
Comment (mkitti) [inline .github/workflows/Format-check.yml:28]:
```suggestion
          julia -e 'using JuliaFormatter; format(["src", "test", "deps", "filters", "gen"], verbose=true)'
```

Comment (t-bltg) [reply .github/workflows/Format-check.yml]: Done.

File: src/attributes.jl
Comment (musm) [reply src/attributes.jl:104]: yep

**Comment by musm** [comment]
can we increase the default line length to 120 or 140? 
I think the following addition is what's needed. 
```
margin = 120
```

**Comment by musm** [comment]
My preference is for yas vs blue style wrt to the alignment of functions and nesting:
https://domluna.github.io/JuliaFormatter.jl/stable/yas_style/#Differences-from-DefaultStyle

**Comment by t-bltg** [comment]
> can we increase the default line length to 120 or 140?

Changed to 120.

> My personal preference is for YAS

Style set to `yas`.

**Comment by mkitti** [comment]
We need to turn off formatting for `gen/apidefs.jl` and `src/api/functions.jl`. I've done this in #1001 .

**Comment by t-bltg** [comment]
The `margin=120` looks odd. I'd stick to the recommended 92 chars margin.

File: src/attributes.jl
Comment (musm) [inline src/attributes.jl:66]: I'm wondering if this should be a one liner or not, since it formats really strangely.

**Comment by musm** [comment]
I think you might be right on it being too long.

File: src/api/functions.jl
Comment (mkitti) [inline src/api/functions.jl:5405]: This is slightly annoying. I'm not sure what the appropriate fix is.

**Comment by t-bltg** [comment]
> I think you might be right on it being too long.

Back to 92.

**Comment by musm** [comment]
Anyone else want to chime in on their preference, as mentioned I think I have a slight preference to the UAS  style.

In general, I'm not crazy about auto-formatters I think sometimes it makes  code less clear than originally written, but I appreciate the consistency.

File: src/api/functions.jl
Comment (t-bltg) [reply src/api/functions.jl:5405]: Probably an editor mistake on my side (not the formatter).

Comment (t-bltg) [reply src/api/functions.jl:5405]: Fixed.

File: src/attributes.jl
Comment (t-bltg) [reply src/attributes.jl:66]: Those are a `yas` effect. Those look better in `blue`.

File: gen/gen_wrappers.jl
Comment (mkitti) [inline gen/gen_wrappers.jl:96]: This indentation looks excessive to me.

File: src/datasets.jl
Comment (mkitti) [inline src/datasets.jl:14]: This looks strange.

File: src/datasets.jl
Comment (mkitti) [inline src/datasets.jl:67]: Too much indentation

File: src/datasets.jl
Comment (mkitti) [inline src/datasets.jl:369]: Too much indentation

File: src/dataspaces.jl
Comment (mkitti) [inline src/dataspaces.jl:83]: Too much indentation

**Comment by t-bltg** [comment]
> In general, I'm not crazy about auto-formatters I think sometimes it makes code less clear than originally written, but I appreciate the consistency.

They are not perfect, but imo the worst you can do is being inconsistent across the repo.

**Comment by musm** [comment]
I'm not against this change, but I disagree in that I don't think inconsistency is the worst you can do. 
Base Julia is inconsistent across the whole repo and the style does vary depending on who wrote what.

**Comment by musm** [comment]
In general, everything looks fine, I'm just debating if we should switch back to blue, some of the changes with YAS are indeed annoying.

**Review by mkitti — COMMENTED** [review]
Some of the indentation looks excessive. I think this is due to YAS.

**Comment by t-bltg** [comment]
> Some of the indentation looks excessive. I think this is due to YAS.

Agreed, back to blue.

**Review by musm — APPROVED** [review]

File: test/readremote.jl
Comment (mkitti) [inline test/readremote.jl:82]: Before was better

File: src/attributes.jl
Comment (mkitti) [reply src/attributes.jl:104]: Let's do this after this pull request

**Comment by t-bltg** [comment]
>  Base Julia is inconsistent across the whole repo and the style does vary depending on who wrote what.

Exactly, `JuliaLang/julia` should have a format auto-check for PRs, and stick to an arbitrary format.
Those are good coding practices everyone should adopt.
Unfortunately, a PR to `julia` would be pointless (0% chance of being merged) given the flame wars it would trigger.

File: src/file.jl
Comment (mkitti) [inline src/file.jl:27]:
```suggestion
    #! format: off
    rd, wr, cr, tr, ff =
```

File: src/file.jl
Comment (mkitti) [inline src/file.jl:33]:
```suggestion
        # mode == "a"  ? (true,  true,  true,  true,  true ) :
        #! format: on
```

File: src/properties.jl
Comment (mkitti) [inline src/properties.jl:318]: I'm confused why it modified this and not some of the others below.

File: src/properties.jl
Comment (mkitti) [inline src/properties.jl:397]: Why did it decide to keep this ternary and not the others?

File: src/typeconversions.jl
Comment (mkitti) [inline src/typeconversions.jl:124]: Alignment as it was before would be good.

**Review by mkitti — COMMENTED** [review]
Closer...

**Comment by mkitti** [comment]
Maybe we should just merge and revert some of this manually. I think we should also add a preformatting tag unless we want to release before merging this.

**Comment by mkitti** [comment]
Ok, I think we're close enough. I'm going to make a few manual changes to turn off formatting in some sections, and then merge.

**Comment by musm** [comment]
> Ok, I think we're close enough. I'm going to make a few manual changes to turn off formatting in some sections, and then merge.

SGTM


Do we need a pre-tag, maybe just the commit is good enough.

File: src/file.jl
Comment (musm) [reply src/file.jl:31]: Yeah, definitely. I think that's the general issue with auto formatting. There's always some bits of code that are worse off.

**Comment by mkitti** [comment]
We just need a easy to remember reference for "before_formatting" in case we find other situations where the formatting got worse.

**Review by mkitti — APPROVED** [review]
Approved pending CI

**Comment by mkitti** [comment]
I am planning to rebase and merge this.

**Comment by musm** [comment]
I'd recommend squash + merge  next time

File: src/file.jl
Comment (t-bltg) [reply src/file.jl]: Tradeoffs :shrug:

**Comment by t-bltg** [comment]
Thanks for the help and valuable comments towards better code quality :100:

**Comment by mkitti** [comment]
> I'd recommend squash + merge next time

I intentionally chose not to squash here since t-bltg explicitly tried to divide up his work into three distinct, atomic commits. Perhaps I should have squashed my commits together.

**Comment by mkitti** [comment]
By the way, it seems that HDF5 upstream also recently started doing automatic formatting using `clang-format`:
https://github.com/HDFGroup/hdf5/blob/develop/.clang-format

**Pull request merged**
